platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/ServiceRegistryBuilder.java

```diff
@@ -52,17 +53,38 @@ public ServiceRegistryBuilder provider(Object provider) {
     /**
      * Providing a scope makes the resulting {@link ServiceRegistry}
      * validate all registered services for being annotated with the given scope.
+     * <p>
+     * However, this still allows to register services without the
+     * {@link org.gradle.internal.service.scopes.ServiceScope @ServiceScope} annotation.
+     *
+     * @see #scopeStrictly(Class)
      */
     public ServiceRegistryBuilder scope(Class<? extends Scope> scope) {
         this.scope = scope;
+        this.strict = false;
+        return this;
+    }
+
+    /**
+     * Providing a scope makes the resulting {@link ServiceRegistry}
+     * validate all registered services for being annotated with the given scope.
+     * <p>
+     * All registered services require the {@link org.gradle.internal.service.scopes.ServiceScope @ServiceScope}
+     * annotation to be present and contain the given scope.
+     *
+     * @see #scope(Class)
+     */
+    public ServiceRegistryBuilder scopeStrictly(Class<? extends Scope> scope) {
```

Comment:
💭 A new internal API to make a given service registry strictly scoped when constructing it via the `ServiceRegistryBuilder`